server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java

```diff
@@ -417,29 +419,27 @@ protected <Request extends ActionRequest, Response extends ActionResponse> void
     }
 
     @Override
-    public Client projectClient(ProjectId projectId) {
+    public ProjectClient projectClient(ProjectId projectId) {
         // We only take the shortcut when the given project ID matches the "current" project ID. If it doesn't, we'll let #executeOnProject
         // take care of error handling.
         if (projectResolver.supportsMultipleProjects() == false && projectId.equals(projectResolver.getProjectId())) {
-            return this;
-        }
-        return new FilterClient(this) {
-            @Override
-            protected <Request extends ActionRequest, Response extends ActionResponse> void doExecute(
-                ActionType<Response> action,
-                Request request,
-                ActionListener<Response> listener
-            ) {
-                projectResolver.executeOnProject(projectId, () -> super.doExecute(action, request, listener));
-            }
-
-            @Override
-            public Client projectClient(ProjectId projectId) {
-                throw new IllegalStateException(
-                    "Unable to create a project client for project [" + projectId + "], nested project client creation is not supported"
-                );
+            // We construct a dedicated project client for the default project if we're in single project mode. This dedicated project
+            // client is an optimization in that it does not use the project resolver and instead executes the request directly.
+            if (defaultProjectClient == null) {
+                defaultProjectClient = new ProjectClient(this, projectId) {
```

Comment:
I wanted to avoid introducing a performance regression in single-project mode by having to create a new client on every invocation, so I went for a non-final field. Other suggestions are welcome too.

Reply:
The issue is that we're constructing the ProjectClient inside the constructor of AbstractClient, but ProjectClient itself also extends AbstractClient. So the place inside the constructor doesn't matter much - if I understand your suggestion correct.

I just realized that we could probably construct the default project client conditionally, only if `this` is not an instance of ProjectClient (as we'll never need a default project client in that case anyway). What do you think of that solution?

Reply:
Btw, it might also worth adding a comment to this field to explain its purpose and reason for creating it conditionally.

Reply:
Hm, calling `projectResolver.supportsMultiProjects()` in the constructor is making _a lot_ of tests fail because they use `TestProjectResolvers.alwaysThrow()`. TBH, I'm more inclined to remove the `supportsMultiProjects()` call instead of updating all those tests. What do you think?

Reply:
Hm ok, and I assume you're also no fan of changing `supportsMultiProjects()` in  `TestProjectResolvers.alwaysThrow()` to return `false` instead of throwing? I'm just trying to find a way to avoid having to change dozens of test classes.

Reply:
Yep unless we rename `alwaysThrow` to something else to reflect its new behaviour. To help with progress here, we can drop the `projectResolver.supportsMultiProjects` check in the constructor in this PR and get back to it in a follow-up? What do you think?